### PR TITLE
feat(elevation_map_loader): add support for differential_map_loader

### PR DESCRIPTION
**Pull request by Shin-kyoto**
## Description

These PRs must be merged before these PRs.
- [PR for autoware_msg](https://github.com/autowarefoundation/autoware_msgs/pull/53)
   - modify service
- [PR for autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/268)
   - modify config for `pointcloud_map_loader`

I would like to add `differential_map_loader` support to `elevation_map_loader`.

In the default setting, `differential_map_loader` support is disabled. So if you want to use this, please change your launcher like [this](https://github.com/Shin-kyoto/autoware.universe/commit/69b1e7b9cc2139858bef8922974b43b698a0e12a).

## Related links

- [TIER IV COMPANY INTERNAL LINK](https://tier4.atlassian.net/browse/T4PB-23166)
- [PR for autoware_msg](https://github.com/autowarefoundation/autoware_msgs/pull/53)
   - modify service
- [PR for autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/268)
   - modify config for `pointcloud_map_loader`

## Tests performed

I tested with a data from Autoware tutorial. The map is divided into 20m grids ([sample-map-rosbag_split.zip](https://github.com/autowarefoundation/autoware.universe/files/10349104/sample-map-rosbag_split.zip)).

- I have confirmed that the Autoware performs the same as the current Autoware when `use_differential_load` is set false.

## Notes for reviewers

How to check

1. Use PRs below
   - [PR for autoware_msg](https://github.com/autowarefoundation/autoware_msgs/pull/53)
   - [PR for autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/268)
2. Build autoware
3. Change your launcher like [this](https://github.com/Shin-kyoto/autoware.universe/commit/69b1e7b9cc2139858bef8922974b43b698a0e12a).
4. Run command below
```bash
ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/data/map/sample-map-rosbag_split pointcloud_map_file:=pointcloud_map vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit planning:=false control:=false system:=false rviz:=true sensing:=false|grep -e elevation -e map
```
5. You should get the outputs like;
```bash
❯ ros2 launch autoware_launch logging_simulator.launch.xml map_path:=$HOME/data/map/sample-map-rosbag_split pointcloud_map_file:=pointcloud_map vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit planning:=false control:=false system:=false rviz:=true sensing:=false|grep -e elevation -e map
........
[component_container-3] [INFO 1679896837.303861415] [map.pointcloud_map_loader]: Load /home/shintarotomie/data/map/sample-map-rosbag_split/pointcloud_map/89400_42300.pcd (1 out of 23)
........
[component_container_mt-1] [INFO 1679896838.782485794] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: send a request to map_loader
[component_container_mt-1] [INFO 1679896838.782555553] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: waiting response
[component_container_mt-1] [INFO 1679896838.792788277] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: concat maps
[component_container_mt-1] [INFO 1679896838.796591044] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: send a request to map_loader
[component_container_mt-1] [INFO 1679896838.796639893] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: waiting response
[component_container_mt-1] [INFO 1679896838.796889053] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: concat maps
[component_container_mt-1] [INFO 1679896838.796905702] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: finish receiving
[component_container_mt-1] [INFO 1679896838.799180362] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: receive service with pointcloud_map
[component_container_mt-1] [INFO 1679896838.799207985] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Create elevation map from pointcloud map 
[component_container_mt-1] [INFO 1679896839.235498083] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Finish creating elevation map. Total time: 0.433 sec
[component_container_mt-1] [INFO 1679896839.666728529] [rosbag2_storage]: Opened database '/home/shintarotomie/work/2023/0317/autoware/install/elevation_map_loader/share/elevation_map_loader/data/elevation_maps/52fe0e6043a473758c9aab905232c887a7374632dabe1099c27b22effcbd93e1/52fe0e6043a473758c9aab905232c887a7374632dabe1099c27b22effcbd93e1_0.db3' for READ_WRITE.
[component_container_mt-1] [INFO 1679896839.672484128] [perception.obstacle_segmentation.elevation_map.elevation_map_loader]: Saving elevation map successful: true
```
6. Check rviz and you should see the elevation_map like;

image

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp [hunk 1]
```diff
@@ -23,6 +23,7 @@ DifferentialMapLoaderModule::DifferentialMapLoaderModule(
     std::bind(
       &DifferentialMapLoaderModule::onServiceGetDifferentialPointCloudMap, this,
       std::placeholders::_1, std::placeholders::_2));
+  map_upper_limit_ = node->declare_parameter<double>("map_upper_limit");
```
Comment (kminoda) [inline map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp:26]: Should this be `double`, not `int` or `long long int`?

File: map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp [hunk 2]
```diff
@@ -63,10 +64,27 @@ bool DifferentialMapLoaderModule::onServiceGetDifferentialPointCloudMap(
   GetDifferentialPointCloudMap::Request::SharedPtr req,
   GetDifferentialPointCloudMap::Response::SharedPtr res)
 {
+  const double map_upper_limit_margin = 10000;
```
Comment (kminoda) [inline map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp:67]: IMO it's better to elaborate a bit more on the meaning of this variable (why the value is 10000)

File: map/map_loader/src/pointcloud_map_loader/utils.cpp [hunk 3]
```diff
@@ -64,6 +64,7 @@ bool sphereAndBoxOverlapExists(
 bool isGridWithinQueriedArea(
   const autoware_map_msgs::msg::AreaInfo area, const PCDFileMetadata metadata)
 {
+  if (area.type == autoware_map_msgs::msg::AreaInfo::ALL_AREA) return true;
   // Currently, the area load only supports spherical area
```
Comment (kminoda) [inline map/map_loader/src/pointcloud_map_loader/utils.cpp:68]: Can you also add 
```
else if (area.type == autoware_map_msgs::msg::AreaInfo::CYLINDER) {
  // code here
}
```

File: perception/elevation_map_loader/src/elevation_map_loader_node.cpp [hunk 4]
```diff
@@ -170,6 +208,63 @@ void ElevationMapLoaderNode::onVectorMap(
...
+        } else {
+          pointcloud_map.width += new_pointcloud_with_id.pointcloud.width;
+          pointcloud_map.row_step += new_pointcloud_with_id.pointcloud.row_step;
+          pointcloud_map.data.insert(
+            pointcloud_map.data.end(), new_pointcloud_with_id.pointcloud.data.begin(),
+            new_pointcloud_with_id.pointcloud.data.end());
+        }
+        cached_ids.push_back(new_pointcloud_with_id.cell_id);
+      }
+    }
```
Comment (kminoda) [inline perception/elevation_map_loader/src/elevation_map_loader_node.cpp:262]: How about this?
```suggestion
    if (result.get()->new_pointcloud_with_ids.empty()) {
      RCLCPP_INFO(this->get_logger(), "finish receiving");
      is_all_received = true;
      break;

    // concat maps
    for (const auto & new_pointcloud_with_id : result.get()->new_pointcloud_with_ids) {
      if (pointcloud_map.width == 0) {
        pointcloud_map = new_pointcloud_with_id.pointcloud;
      } else {
        pointcloud_map.width += new_pointcloud_with_id.pointcloud.width;
        pointcloud_map.row_step += new_pointcloud_with_id.pointcloud.row_step;
        pointcloud_map.data.insert(
          pointcloud_map.data.end(), new_pointcloud_with_id.pointcloud.data.begin(),
          new_pointcloud_with_id.pointcloud.data.end());
      }
      cached_ids.push_back(new_pointcloud_with_id.cell_id);
    }
```

**Review by kminoda — COMMENTED** [review]
Thank you! I left some comment :pray:

File: map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp [hunk 5]
```diff
@@ -63,10 +64,27 @@ bool DifferentialMapLoaderModule::onServiceGetDifferentialPointCloudMap(
   GetDifferentialPointCloudMap::Request::SharedPtr req,
   GetDifferentialPointCloudMap::Response::SharedPtr res)
 {
+  const int64_t map_upper_limit_margin = 10000;
```
Comment (yukkysaito) [inline map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp:67]: what is the unit? (size?)

File: map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp [hunk 6]
```diff
@@ -63,10 +64,27 @@ bool DifferentialMapLoaderModule::onServiceGetDifferentialPointCloudMap(
...
   auto area = req->area;
   std::vector<std::string> cached_ids = req->cached_ids;
   differentialAreaLoad(area, cached_ids, res);
   res->header.frame_id = "map";
+  // calculate pcd map size to respond
+  double response_pcd_size = 0;
```
Comment (yukkysaito) [inline map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp:73]: Is there a reason why you use double even though it is size?

File: perception/elevation_map_loader/src/elevation_map_loader_node.cpp [hunk 7]
```diff
@@ -128,6 +151,21 @@ void ElevationMapLoaderNode::publish()
     pcl::toROSMsg(*elevation_map_cloud_ptr, elevation_map_cloud_msg);
     pub_elevation_map_cloud_->publish(elevation_map_cloud_msg);
   }
+  is_elevation_map_published_ = true;
+}
+
+void ElevationMapLoaderNode::timerCallback()
+{
+  // flag to make receiveMap() called only once.
+  static bool is_map_received = false;
```
Comment (yukkysaito) [inline perception/elevation_map_loader/src/elevation_map_loader_node.cpp:160]: I think better to put it as a member variable.

File: map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp (hunk 2 above)
Comment (Shin-kyoto) [reply map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp]: I removed map_upper_limit_margin and added description in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/ae4e410edb2f66f34d20cd4d07a47e8d5d8b9124

File: map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp (hunk 5 above)
Comment (Shin-kyoto) [reply map/map_loader/src/pointcloud_map_loader/differential_map_loader_module.cpp]: I removed map_upper_limit_margin and added description in https://github.com/autowarefoundation/autoware.universe/commit/ae4e410edb2f66f34d20cd4d07a47e8d5d8b9124

File: map/map_loader/src/pointcloud_map_loader/utils.cpp (hunk 3 above)
Comment (Shin-kyoto) [reply map/map_loader/src/pointcloud_map_loader/utils.cpp]: Add conditional branch for CYLINDER in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/ae4e410edb2f66f34d20cd4d07a47e8d5d8b9124

File: map/map_loader/src/pointcloud_map_loader/utils.cpp [hunk 8]
```diff
@@ -64,10 +64,14 @@ bool sphereAndBoxOverlapExists(
...
 {
-  if (area.type == autoware_map_msgs::msg::AreaInfo::ALL_AREA) return true;
-  // Currently, the area load only supports spherical area
-  geometry_msgs::msg::Point center = area.center;
-  double radius = area.radius;
-  bool res = sphereAndBoxOverlapExists(center, radius, metadata.min, metadata.max);
+  bool res = false;
+  if (area.type == autoware_map_msgs::msg::AreaInfo::ALL_AREA) {
+    res = true;
+  } else if (area.type == autoware_map_msgs::msg::AreaInfo::CYLINDER) {
```
Comment (kminoda) [inline map/map_loader/src/pointcloud_map_loader/utils.cpp:70]: My bad, this was not cylinder but a sphere. Please fix this (autoware_map_msgs as well)

File: perception/elevation_map_loader/src/elevation_map_loader_node.cpp (hunk 7 above)
Comment (Shin-kyoto) [reply perception/elevation_map_loader/src/elevation_map_loader_node.cpp]: Apply it in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/5ee3685f8f63f3b0183f4dda49f447e971cc39d6

File: perception/elevation_map_loader/src/elevation_map_loader_node.cpp [hunk 9]
```diff
@@ -79,11 +80,33 @@ ElevationMapLoaderNode::ElevationMapLoaderNode(const rclcpp::NodeOptions & optio
...
     "/api/autoware/get/map/info/hash", durable_qos,
     std::bind(&ElevationMapLoaderNode::onMapHash, this, _1));
-  sub_pointcloud_map_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
-    "input/pointcloud_map", durable_qos,
-    std::bind(&ElevationMapLoaderNode::onPointcloudMap, this, _1));
   sub_vector_map_ = this->create_subscription<autoware_auto_mapping_msgs::msg::HADMapBin>(
     "input/vector_map", durable_qos, std::bind(&ElevationMapLoaderNode::onVectorMap, this, _1));
+  if (use_differential_load) {
+    {
+      const auto period_ns =
```
[the remark this answers is not in the document]
Comment (Shin-kyoto) [reply perception/elevation_map_loader/src/elevation_map_loader_node.cpp]: Apply it in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/83e834a049ed0486a2fe93df21597782013ab807

File: map/map_loader/src/pointcloud_map_loader/utils.cpp (hunk 8 above)
Comment (Shin-kyoto) [reply map/map_loader/src/pointcloud_map_loader/utils.cpp]: Thanks!
Correct CYLINDER to SPHERE in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/64b63d1bf9195c095b3c77e22da2762210db6e50

File: perception/elevation_map_loader/src/elevation_map_loader_node.cpp (hunk 4 above)
Comment (Shin-kyoto) [reply perception/elevation_map_loader/src/elevation_map_loader_node.cpp:261]: Thank you!
However, IMO, flag(`is_all_received`) is more suitable than `break` for readability.
What do you think about it?

File: map/map_loader/README.md [hunk 10]
```diff
@@ -38,13 +38,14 @@ Please see [the description of `GetDifferentialPointCloudMap.srv`](https://githu
...
+| enable_differential_load      | bool    | A flag to enable differential pointcloud map server                                                                                                                                                                                                                                                                                 | false         |
+| leaf_size                     | float   | Downsampling leaf size (only used when enable_downsampled_whole_load is set true)                                                                                                                                                                                                                                                   | 3.0           |
+| map_upper_limit               | int64_t | Limit data size of topic/service/action (determined by cyclonedds) (only used when differential_load enabled). Default value is calculated as follows; <br> map_upper_limit = floor((maximum receive buffer size)\* 0.8) <br> If maximum receive buffer size in your PC's environment is 2147483647, map_upper_limit is 1717986917. | 1717986917    |
```
Comment (kminoda) [inline map/map_loader/README.md:48]: How about clarifying where this magic number come from? (probably here? https://github.com/eclipse-cyclonedds/cyclonedds/blob/09a02f5b2b9ba856d7d4c616510252cf9e508ec2/docs/manual/config/config_file_reference.rst#cycloneddsdomaininternalmaxsamplesize)

Comment (kminoda) [inline map/map_loader/README.md:48]: Can you also add a description on how map_upper_limit works? (What happens if the data size exceeds map_upper_limit?)

Comment (Shin-kyoto) [reply map/map_loader/README.md]: Thank you!!
I added description in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/56352bbc57b84f00fdcba9fd14144d12812da67d

Comment (Shin-kyoto) [reply map/map_loader/README.md]: Thank you!!
I added description in https://github.com/autowarefoundation/autoware.universe/commit/56352bbc57b84f00fdcba9fd14144d12812da67d

File: perception/elevation_map_loader/src/elevation_map_loader_node.cpp (hunk 4 above)
Comment (Shin-kyoto) [reply perception/elevation_map_loader/src/elevation_map_loader_node.cpp]: Thank you for your review!
I added function concatenating pointcloud maps in https://github.com/autowarefoundation/autoware.universe/pull/3162/commits/ebe59189c85bfa6e9daac52f7c7b24a22fcc65bc, and now nested complexity depth decreased.

**Pull request closed without merging**